warehouse/search/__init__.py

```diff
@@ -18,7 +18,32 @@
 from celery.schedules import crontab
 from elasticsearch_dsl import serializer
 
+from warehouse import db
 from warehouse.search.utils import get_index
+from warehouse.packaging.models import Project, Release
+
+
+@db.listens_for(db.Session, "after_flush")
+def store_projects_for_project_reindex(config, session, flush_context):
+    # We'll (ab)use the session.info dictionary to store a list of pending
+    # Project updates to the session.
+    purges = session.info.setdefault("warehouse.search.project_updates", set())
+
+    # Go through each new, changed, and deleted object and attempt to store
+    # a Project to reindex for when the session has been committed.
+    for obj in (session.new | session.dirty | session.deleted):
+        if obj.__class__ == Project:
+            purges.add(obj)
```

Comment:
If the project has been deleted, I'm not sure it makes sense to add it to the set here, because the `.filter` below will not return any releases at all. (I guess a deleted project can just fall out of the index when we reindex nightly instead of trying to do something special for it here.)

However if a release has been deleted, it _does_ make sense.